### DISCUSSION TITLE
Commit Container?

### DISCUSSION THREAD

**Discussion by schmitch** [testcontainers/testcontainers-dotnet#1072] [Q&A] [answered]
Hello, I'm currently looking into testcontainers to speedup our test process.
Actually the biggest slowdown comes from loading the migrations. thats why we use respawn, however that makes it tough when we are dealing with multiple test assemblies, because each assembly would use its own database and thus each assembly has the same slowdown 4-5 seconds just to load all migrations initally. so I tought of using testcontainers to use it like that:

```
            var localContainer = new PostgreSqlBuilder()
                .WithDatabase("asvg_salestool")
                .WithUsername("asvg_salestool")
                .Build();
            await localContainer.StartAsync();
            
            Console.WriteLine($"Container inside the mutex {localContainer.Id}");

            await localContainer.ExecScriptAsync(await File.ReadAllTextAsync(script));
```

and now I want to commit the container so that it can be used by all test assemblies. I can build this container either with a msbuild task or with a `Mutex` and than I can just check if the image exists and thus it would be hella fast and I do not need to deal with respawn at all.
Is there any way to commit the container to an image like `docker commit`?

I know that I can put the sql script inside the container but the script would still load a while, so I want to load it upfront and save it so that postgres always has the initial migration data already inside the container.

**Comment by HofmeisterAn** [comment]
Committing a container state is not yet built into Testcontainers for .NET. I double-checked the Docker Engine API, including Docker.DotNet, both expose the necessary APIs 👍. This is something we need to implement. Would you be interested in contributing this feature? In the meantime, you can achieve something similar with the image builder API (schema initialization during the build) or by utilizing Docker volumes. However, both approaches will not work with the integrated cleanup mechanism of Testcontainers, so you will need to disable it.

**Reply by schmitch** [reply]
I might be able to look into it but I'm not sure if I might find time the next weeks or so 😭

**Reply by HofmeisterAn** [reply]
FYI: This issue seems to relate: https://github.com/testcontainers/testcontainers-dotnet/issues/1076.

**Reply by schmitch** [reply]
I actually tried with Docker.DotNet and it looks that it will cleanup resources afterwards, which well is strange.
basically I did that:

```
            await localContainer.ExecScriptAsync(script);
            await localContainer.StopAsync();

            var containerId = localContainer.Id;
            var response = await client.Images.CommitContainerChangesAsync(
                new CommitContainerChangesParameters
                {
                    ContainerID = containerId,
                    RepositoryName = "docker.envisia.io/test-asvg",
                    Tag = "latest",
                }
            );

```

I also used `images = await client.Images.ListImagesAsync(new ImagesListParameters { });` to show images that it build and it build the container.
I have no idea why it cleans it up. it's also not working, probably because of another reason, probably because of the fsync off stuff.

And also it cleans it too fast, because I want to keep the image until ALL proceses are done, so maybe I really need to do it differently 😢  use Docker.DotNet directly and create a msbuild task probably...

**Reply by HofmeisterAn** [reply, marked as answer]
> it will cleanup resources afterwards, which well is strange.

> I have no idea why it cleans it up.

This is very likely, [Ryuk](https://dotnet.testcontainers.org/api/resource-reaper/). It tracks resources and cleans them up after the test session/run to prevent the system from getting littered. I do not recommend disabling Ryuk. Instead, we should figure out how to remove the label from the committed image. Let's continue the discussion on the mentioned issue.
